File: core/security/otp_privacy.py
"""
Sensitive Data & OTP Privacy Engine (Python)
Implements detection & strict pipeline termination rules for OTP, PIN, CVV,
passwords, and banking verification codes.
"""

import re
from typing import Dict, Any, Optional

OTP_PATTERNS = [
    re.compile(r"\b(?:otp|one[- ]time[- ]password|verification\s*code|auth\s*code|security\s*code)\b(?:\s+is\s+|[:\s-])*(\d{4,8})\b", re.IGNORECASE),
    re.compile(r"\b(\d{4,8})\s+(?:is\s+your\s+(?:otp|verification|security|login)\s+code)\b", re.IGNORECASE),
    re.compile(r"\b(?:pin|upi[- ]pin|mpin)\b(?:\s+is\s+|[:\s-])*(\d{4,6})\b", re.IGNORECASE),
    re.compile(r"\b(?:cvv|cvc)\b(?:\s+is\s+|[:\s-])*(\d{3,4})\b", re.IGNORECASE),
    re.compile(r"\b(?:password|passcode)\b(?:\s+is\s+|[:\s-])*([^\s,]{6,})\b", re.IGNORECASE),
]


class OtpPrivacyEngine:
    @staticmethod
    def evaluate_content(text: str) -> Dict[str, Any]:
        """Evaluates incoming text for sensitive authentication items."""
        if not text or not isinstance(text, str):
            return {
                "is_sensitive": False,
                "category": "NONE",
                "should_abort_pipeline": False,
                "suppress_audio": False,
                "suppress_storage": False,
            }

        for pattern in OTP_PATTERNS:
            match = pattern.search(text)
            if match:
                lower = text.lower()
                category = "OTP"
                if "otp" in lower or "one-time" in lower or "one time" in lower:
                    category = "OTP"
                elif "cvv" in lower or "cvc" in lower:
                    category = "CVV"
                elif "upi" in lower or "mpin" in lower:
                    category = "UPI_PIN"
                elif "pin" in lower:
                    category = "PIN"
                elif "password" in lower or "passcode" in lower:
                    category = "PASSWORD"
                elif "bank" in lower:
                    category = "BANKING_CODE"

                return {
                    "is_sensitive": True,
                    "category": category,
                    "classification": "OTP" if category == "OTP" else ("SECURITY_CODE" if category == "BANKING_CODE" else ("FINANCIAL" if category in ("PIN", "UPI_PIN", "CVV") else "CREDENTIAL")),
                    "action": "BLOCK",
                    "should_abort_pipeline": True,
                    "suppress_audio": True,
                    "suppress_storage": True,
                    "sanitized_snippet": "[OTP REDACTED]" if category == "OTP" else "[SENSITIVE_CODE_REDACTED]",
                    "safe_content": "[OTP REDACTED]" if category == "OTP" else "[REDACTED]",
                }

        return {
            "is_sensitive": False,
            "category": "NONE",
            "classification": "NORMAL",
            "action": "ALLOW",
            "should_abort_pipeline": False,
            "suppress_audio": False,
            "suppress_storage": False,
            "safe_content": text,
        }

    @staticmethod
    def purge_temporary_buffer(buffer_id: str) -> None:
        """Purges buffered notification data upon sensitive detection (Phase 2 hook)."""
        pass

File: core/security/otpPrivacy.ts
/**
 * Sensitive Data & OTP Privacy Engine (Phase 2 - Step 4 Upgrade)
 * Enforces the strict privacy rule:
 * Notification -> Sensitive Data Detector -> Classification -> Policy Decision
 *
 * If OTP/security code is detected:
 * - immediately stop notification processing
 * - do not speak it
 * - do not send the actual OTP to an AI model
 * - do not store it in memory
 * - delete the temporary notification content
 * - do not open the application
 * - lock/suppress notification reading
 */

import { SensitiveDataDetector } from './sensitiveDataDetector.ts';
import { PrivacyPolicy } from './privacyPolicy.ts';
import { OtpLock } from './otpLock.ts';
import { SecurityClassification, PrivacyAction } from './types.ts';

export interface SensitiveDetectionResultLegacy {
  isSensitive: boolean;
  category: 'OTP' | 'PIN' | 'CVV' | 'PASSWORD' | 'UPI_PIN' | 'BANKING_CODE' | 'NONE';
  classification: SecurityClassification;
  action: PrivacyAction;
  shouldAbortPipeline: boolean;
  suppressAudio: boolean;
  suppressStorage: boolean;
  sanitizedSnippet?: string;
  safeContent: string;
}

export class OtpPrivacyEngine {
  /**
   * Evaluates text from incoming notifications or messages.
   * If sensitive security data is detected, triggers strict abort policy.
   */
  public static evaluateContent(text: string): SensitiveDetectionResultLegacy {
    if (!text || typeof text !== 'string') {
      return {
        isSensitive: false,
        category: 'NONE',
        classification: 'NORMAL',
        action: 'ALLOW',
        shouldAbortPipeline: false,
        suppressAudio: false,
        suppressStorage: false,
        safeContent: '',
      };
    }

    const detection = SensitiveDataDetector.detect(text);
    const policy = PrivacyPolicy.evaluate(detection, text);

    let legacyCategory: SensitiveDetectionResultLegacy['category'] = 'NONE';
    if (detection.classification === 'OTP') legacyCategory = 'OTP';
    else if (detection.detectedTypes.includes('UPI_PIN')) legacyCategory = 'UPI_PIN';
    else if (detection.detectedTypes.includes('CVV')) legacyCategory = 'CVV';
    else if (detection.detectedTypes.includes('PIN')) legacyCategory = 'PIN';
    else if (detection.detectedTypes.includes('PASSWORD')) legacyCategory = 'PASSWORD';
    else if (detection.classification === 'SECURITY_CODE') legacyCategory = 'BANKING_CODE';

    if (detection.classification === 'OTP') {
      OtpLock.enforce(text);
      return {
        isSensitive: true,
        category: 'OTP',
        classification: 'OTP',
        action: 'BLOCK',
        shouldAbortPipeline: true,
        suppressAudio: true,
        suppressStorage: true,
        sanitizedSnippet: '[OTP REDACTED]',
        safeContent: '[OTP REDACTED]',
      };
    }

    return {
      isSensitive: detection.isSensitive,
      category: legacyCategory,
      classification: detection.classification,
      action: policy.action,
      shouldAbortPipeline: policy.shouldAbortPipeline,
      suppressAudio: policy.suppressAudio,
      suppressStorage: policy.suppressStorage,
      sanitizedSnippet: detection.isSensitive ? '[SENSITIVE_CODE_REDACTED]' : undefined,
      safeContent: policy.action === 'BLOCK' ? '[REDACTED]' : text,
    };
  }

  /**
   * Purges temporary notification buffer if sensitive.
   */
  public static purgeTemporaryBuffer(bufferId: string): void {
    // Purges temporary memory buffer for privacy compliance
  }
}
